src/com/facebook/buck/go/GoDescriptors.java

```diff
@@ -127,4 +147,84 @@ static GoLibrary createGoLibraryRule(BuildRuleParams params, BuildRuleResolver r
         goBuckConfig.getGoCompiler().get());
 
   }
+
+  static GoBinary createGoBinaryRule(BuildRuleParams params, BuildRuleResolver resolver,
+      GoBuckConfig goBuckConfig, CxxPlatform cxxPlatform, ImmutableSet<SourcePath> srcs,
+      List<String> compilerFlags, List<String> linkerFlags) {
+    BuildTarget libraryTarget =
+        BuildTarget.builder(params.getBuildTarget())
+            .addFlavors(ImmutableFlavor.of("compile"))
+            .build();
+    GoLibrary library = GoDescriptors.createGoLibraryRule(
+        params.copyWithBuildTarget(libraryTarget),
+        resolver,
+        goBuckConfig,
+        Paths.get("main"),
+        srcs,
+        compilerFlags
+    );
+    resolver.addToIndex(library);
+
+    BuildRuleParams binaryParams = params.copyWithDeps(
+        Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of(library)),
+        Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of())
+    );
+
+    GoSymlinkTree symlinkTree = GoDescriptors.requireTransitiveSymlinkTreeRule(
+        binaryParams,
+        resolver);
+
+    return new GoBinary(
+        params.copyWithDeps(
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of(symlinkTree, library)),
+            Suppliers.ofInstance(ImmutableSortedSet.<BuildRule>of())),
+        new SourcePathResolver(resolver),
+        cxxPlatform.getLd(),
+        symlinkTree,
+        library,
+        goBuckConfig.getGoLinker().get(),
+        ImmutableList.<String>builder()
+            .addAll(goBuckConfig.getLinkerFlags())
+            .addAll(linkerFlags)
+            .build());
+  }
+
+  static synchronized Tool getTestMainGenerator(
+      GoBuckConfig goBuckConfig, BuildRuleResolver resolver, CxxPlatform cxxPlatform,
+      ProjectFilesystem projectFilesystem, Function<Optional<String>, Path> cellRoots,
+      RuleKeyBuilderFactory ruleKeyBuilderFactory) {
+    Optional<Tool> configTool = goBuckConfig.getGoTestMainGenerator(resolver);
+    if (configTool.isPresent()) {
+      return configTool.get();
+    }
+
+    BuildTarget testGenTarget = BuildTarget.builder(
```

Comment:
Ah, I see, we don't have a namespace reserved for generated targets, but maybe we should.

One thing is that it's a bit confusing to have a target show up out of nowhere.

@shs96c Any opinions on this?